### PR TITLE
Ex06

### PR DESCRIPTION
**Pull request by jkorsvik**
This was one of the more troublesome task by far. It runs ok now, would love feedback on how I could improve the code 💯

### CODE REVIEW

File: src/jon_mikkel_ex/ex06/logistic_regression.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+    if type(z) is int or float:
+        return 1 / (1 + np.exp(-z))
+    if type(z) is np.ndarray:
+        sigmoidal_transformed_z = np.zeros_like(z, np.float64)
+        for ind, val in enumerate(z):
+            sigmoidal_transformed_z[ind] = 1 / (1 + np.exp(-val))
+        return sigmoidal_transformed_z
+    else:
+        raise TypeError('Z input has to be a np.ndarray, alternatively\
+ a float or an int will suffice')
```
Comment (danielambrosius) [inline src/jon_mikkel_ex/ex06/logistic_regression.py]: This trick here is that numpy arrays can be added together just like numbers. All operations are then performed element-wise. A valid solution for this function would be `return 1/(1+np.exp(-z))`. This should work with any np.ndarray.

Comment (danielambrosius) [reply src/jon_mikkel_ex/ex06/logistic_regression.py]: Except of course special operations such as `@`, `.` etc.

File: src/jon_mikkel_ex/ex06/logistic_regression.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        The true class labels for each data point.
+
+    Returns
+    -------
+    gradient : np.ndarray(shape=(r,))
+        The gradient of the cross entropy loss related to the linear
+        logistic regression model.
+    """
+    proba = predict_proba(coef, X)
+    return X.T@(y - proba)
```
Comment (danielambrosius) [inline src/jon_mikkel_ex/ex06/logistic_regression.py]: if you experienced a bad accuracy while running the `__main__` section, this might be why. According to https://inf200-coursework.readthedocs.io/en/latest/, it should be `(y_hat - y)` not the opposite (as was written in the script). But it looks like you have accounted for it in line 228, so it should still converge.

File: src/jon_mikkel_ex/ex06/logistic_regression.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+        Returns
+        -------
+        coef : np.ndarray(shape=(n,))
+            The logistic regression weights
+        """
+        for _ in range(self.max_iter):
+            coef += (self.learning_rate * logistic_gradient(coef, X, y))
+            if self._has_converged(coef, X, y):
+                break
+        return coef  # Returns the first coefficient that converges enough
```
Comment (danielambrosius) [inline src/jon_mikkel_ex/ex06/logistic_regression.py]: I would recommend using a while loop here with a double control statement:
```python
while _iter <= learning rate and not _has_converged():
iter += 1
...
```

Comment (danielambrosius) [reply src/jon_mikkel_ex/ex06/logistic_regression.py]: If the function would be longer, its easier to read if the exit statement is in the beginning rather than having to look for a `break` somewhere in the function.

**Review by danielambrosius — APPROVED** [review]
Looks good dude. Approved. Not that many things to improve other than general coding practice. Happy holidays! And remember that there is still a tutorial session on friday 8.
